File: PAT_Basic_Level_by_go/1047 编程团体赛/1047 编程团体赛.go
/*
题目：1047 编程团体赛
实现原理：按队号在哈希表中累加个人分数，并同步记录最高分队伍。
*/
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

// 编程团体赛：统计各队总分，输出冠军队伍及其总分。
func main() {
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Buffer(make([]byte, 1024*1024), 1024*1024)
	if !scanner.Scan() {
		return
	}
	N, _ := strconv.Atoi(strings.TrimSpace(scanner.Text()))
	scores := make(map[int]int)
	maxTeam, maxScore := 0, -1
	for i := 0; i < N; i++ {
		if !scanner.Scan() {
			break
		}
		f := strings.Fields(scanner.Text())
		teamText := strings.SplitN(f[0], "-", 2)[0]
		team, _ := strconv.Atoi(teamText)
		score, _ := strconv.Atoi(f[1])
		scores[team] += score
		if scores[team] > maxScore {
			maxScore = scores[team]
			maxTeam = team
		}
	}
	fmt.Printf("%d %d\n", maxTeam, maxScore)
}

File: PAT_Basic_Level_by_c/1047 编程团体赛/1047 编程团体赛.c
/*
 * 题目：1047 编程团体赛
 * 实现原理：
 *   统计各团队的总分，找出总分最高的团队。
 *   输入格式：团队编号-队员编号 分数
 *   算法步骤：
 *   1. 使用score数组存储每个团队的总分（团队编号作为下标）
 *   2. 解析输入，提取团队编号和分数，累加到对应位置
 *   3. 遍历数组找出最大总分及对应的团队编号
 *   注意：团队编号范围为1-1000，数组大小设为1001即可。
 */
#include <stdio.h>

int main() {
    int N;
    int score[1001] = {0};  // 以团队编号为下标的团队总分数组
    
    scanf("%d", &N);
    for (int i = 0; i < N; i++) {
        int team, member, s;
        scanf("%d-%d %d", &team, &member, &s);  // 输入格式：团队编号-队员编号 分数
        score[team] += s;  // 累加到对应团队
    }
    
    int max_score = 0, max_team = 0;  // 最高总分、对应团队编号
    for (int i = 1; i <= 1000; i++) {
        if (score[i] > max_score) {  // 找到更大总分则更新
            max_score = score[i];
            max_team = i;
        }
    }
    
    printf("%d %d\n", max_team, max_score);
    
    return 0;
}